bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/internal/serviceregistry/ServiceRegistry.java

```diff
@@ -1077,6 +1077,11 @@ private List<ServiceRegistrationImpl<?>> lookupServiceRegistrations(String clazz
 				if (filter instanceof FilterImpl) {
 					// check if we can determine the clazz from the filter
 					String filterObjectClazz = ((FilterImpl) filter).getRequiredObjectClass();
+					if (filterObjectClazz != null && ((FilterImpl) filter).getChildren().isEmpty()) {
+						// this is a simple (objectClass=serviceClass) filter;
+						// no need to evaluate the filter
+						filter = null;
+					}
 					result = filterObjectClazz == null ? allPublishedServices
 							: publishedServicesByClass.get(filterObjectClazz);
 				} else {
```

Comment:
```suggestion
					if (filterObjectClazz != null) {
					  result = publishedServicesByClass.get(filterObjectClazz);
					  if ((FilterImpl) filter).getChildren().isEmpty()) {
						// this is a simple (objectClass=serviceClass) filter;
						// no need to evaluate the filter
						filter = null;
					  }
					} else {
					  result = allPublishedServices;
					}
				} else {
```

Maybe simplify by combining the two filterObjectClazz null tests?